File: 2020/pp-mar 14/lecture_006_TAS/QuickSort.java
import java.util.Random;

public class QuickSort {

    public static Random rand = new Random();

    public static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static int partitionOverPivot(int[] arr, int si, int ei, int pIdx) {
        swap(arr, pIdx, ei);

        int p = si - 1, itr = si;
        while (itr <= ei) {
            if (arr[itr] <= arr[ei])
                swap(arr, itr, ++p);
            itr++;
        }

        return p;
    }

    public static void quickSort(int[] arr, int si, int ei) {
        if (si >= ei)
            return;

        int pIdx = ei; // rand.nextInt(ei - si + 1) + si
        int p = partitionOverPivot(arr, si, ei, pIdx);

        // int prev = si, idx = si;
        // while (idx <= ei && arr[prev] <= arr[idx]) {
        // prev = idx;
        // idx++;
        // }

        // if (idx == ei + 1)
        // return;

        quickSort(arr, si, p - 1);
        quickSort(arr, p + 1, ei);
    }

    public static void main(String[] args) {
        int[] arr = { -12, 2, 7, 4, 34, 23, 0, 1, -1, -50, 16, 23, 7, 4, 2, 3 };
        quickSort(arr, 0, arr.length - 1);

        for (int ele : arr)
            System.out.print(ele + " ");
    }
}

File: 2019/LevelUpDecBatch/l010_SearchingAndSorting/quickSort.cpp
#include <iostream>
#include <vector>

using namespace std;

int segregateData(vector<int> &arr, int si, int ei, int pivot)
{
    swap(arr[pivot], arr[ei]);
    int p = si - 1, itr = si;
    while (itr <= ei)
    {
        if (arr[itr] <= arr[ei])
            swap(arr[++p], arr[itr]);
        itr++;
    }

    return p;
}

void quickSort(vector<int> &arr, int si, int ei)
{
    if (si > ei)
        return;

    // int len = (ei - si + 1);
    // int pivot = rand() % len + si;

    int pivot = ei;
    int pIdx = segregateData(arr, si, ei, pivot);

    //Sorted Array Check
    // if it is sorted then return

    quickSort(arr, si, pIdx - 1);
    quickSort(arr, pIdx + 1, ei);
}



int main()
{
    vector<int> arr = {-12, 2, 7, 4, 34, 23, 0, 1, -1, -50, 16, 23, 7, 4, 2, 3};
    quickSort(arr, 0, arr.size() - 1);

    for (int ele : arr)
        cout << ele << " ";
}
